src/ast/passes/codegen_llvm.cpp

```diff
@@ -3528,11 +3529,200 @@ void CodegenLLVM::createPrintNonMapCall(Call &call, int &id)
+  {
+    elems.push_back(b_.getInt8Ty()->getPointerTo());
+    elems.push_back(b_.getInt8Ty()->getPointerTo());
+  }
+  auto type = StructType::create(elems, "struct map_t", false);
+
+  auto var = llvm::dyn_cast<GlobalVariable>(
+      module_->getOrInsertGlobal(var_name, type));
+  var->setInitializer(ConstantAggregateZero::get(type));
+  var->setSection(".maps");
+  var->setDSOLocal(true);
+  var->addDebugInfo(debuginfo);
+}
+
+// Emit maps in libbpf format so that Clang can create BTF info for them which
+// can be read and used by libbpf.
+//
+// Each map should be defined by a global variable of a struct type with the
+// following fields:
+// - "type"        map type (e.g. BPF_MAP_TYPE_HASH)
+// - "max_entries" maximum number of entries
+// - "key"         key type
+// - "value"       value type
+//
+// "type" and "max_entries" are integers but they must be represented as
+// pointers to an array of ints whose dimension defines the specified value.
+//
+// "key" and "value" are pointers to the corresponding types. Note that these
+// are not used for the BPF_MAP_TYPE_RINGBUF map type.
+//
+// The most important part is to generate BTF with the above information. This
+// is done by emitting DWARF which LLVM will convert into BTF. The LLVM type of
+// the global variable itself is not important, it can simply be a struct with 4
+// pointers.
+//
+// Note that LLVM will generate BTF which misses some information. This is
+// normally set by libbpf's linker but since we load BTF directly, we must do
+// the fixing ourselves, until we start loading BPF programs via bpf_object.
+// See BpfBytecode::fixupBTF for details.
+void CodegenLLVM::generate_maps(const RequiredResources &resources)
+{
+  // User-defined maps
+  for (auto &map_val : resources.map_vals)
+  {
+    const std::string &name = map_val.first;
+
+    SizedType val_type = map_val.second;
+    auto key = resources.map_keys.find(name);
+    if (key == resources.map_keys.end())
+      LOG(FATAL) << "map key for \"" << name << "\" not found";
+
+    auto max_entries = bpftrace_.config_.get(ConfigKeyInt::max_map_keys);
+    auto map_type = libbpf::BPF_MAP_TYPE_UNSPEC;
+    if (val_type.IsCountTy() && key->second.args_.empty())
+    {
+      max_entries = 1;
+      map_type = libbpf::BPF_MAP_TYPE_PERCPU_ARRAY;
+    }
+    else if (bpftrace_.feature_->has_map_percpu_hash() &&
+             (val_type.IsHistTy() || val_type.IsLhistTy() ||
+              val_type.IsCountTy() || val_type.IsSumTy() ||
+              val_type.IsMinTy() || val_type.IsMaxTy() || val_type.IsAvgTy() ||
+              val_type.IsStatsTy()))
+    {
+      map_type = libbpf::BPF_MAP_TYPE_PERCPU_HASH;
+    }
+    else
+    {
+      map_type = libbpf::BPF_MAP_TYPE_HASH;
+    }
+
+    createMapDefinition(name, map_type, max_entries, key->second, val_type);
+  }
+
+  // bpftrace internal maps
+  for (const StackType &stack_type : resources.stackid_maps)
+  {
+    // bpf_get_stackid() is kind of broken by design.
```

Comment:
@viktormalik I had it on my internal todos to replace get_stackid but forgot to write up an issue until just now :) 

I stole some of your text but at least now you can replace this large comment with just a link to the issue:
https://github.com/iovisor/bpftrace/issues/2962

Reply:
Thanks for creating the issue! The text is not mine, I just copied it from other place (where it will be removed in one of the future PRs).

Reply:
Good idea on linking the issue, though, will do that!